File: #139_Word_Break/Solution.java
import java.util.Set;
import java.util.HashSet;

public class Solution {
    public boolean wordBreak(String s, Set<String> wordDict) {
        if (s == null || s.length() == 0) return false;
        if (wordDict == null || wordDict.isEmpty()) return false;
        int len = s.length();
        boolean[] dp = new boolean[len + 1];
        dp[0] = true;
        for (int i = 0; i < len; i++) {
        	for (int k = 0; k <= i; k++) {
        		if (dp[k] && wordDict.contains(s.substring(k, i + 1))) {
        			dp[i + 1] = true;
        			break;
        		}
        	}
        }
        return dp[len];        
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	String s = "leetcode";
    	Set<String> set = new HashSet<String>();
    	set.add("leet"); set.add("code");
    	System.out.println(sol.wordBreak(s, set));
    }
}

File: #139_Word_Break/Solution.py
'''
O(n ^ 2) time Solution
'''

class Solution:
    # @param s, a string
    # @param wordDict, a set<string>
    # @return a boolean
    def wordBreak(self, s, wordDict):
        if not wordDict:
        	return False
        dp = [False for x in range(len(s) + 1)]
        dp[0] = True
        for j in range(len(s)):
        	for i in range(j + 1):
        		if s[i : j + 1] in wordDict and dp[i]:
        			dp[j + 1] = True
        return dp[len(s)]

sol = Solution()
wd = set(['leet', 'code'])
print sol.wordBreak('leetcode', wd)